controller_connections.go

```diff
@@ -90,19 +92,116 @@ func (c *controller) handleIncoming(con net.Conn) {
 		return
 	}
 
-	peer := newPeer(c.net, c.peerStatus, c.peerData)
-	// we are never expecting a reject-alternate for incoming connections
-	if _, err := peer.StartWithHandshake(ep, con, true); err != nil {
-		c.logger.WithError(err).Debugf("Handshake failed for address %s, stopping", ep)
-		peer.Stop()
+	// don't bother with alternatives for incoming connections
+	peer, _, err := c.handshake(host, con, true)
+
+	if err != nil {
+		c.logger.WithError(err).Debugf("inbound connection from %s failed handshake", host)
+		con.Close()
 		return
 	}
 
 	c.logger.Debugf("Incoming handshake success for peer %s, version %s", peer.Hash, peer.prot.Version())
+}
+
+func (c *controller) handshake(ip string, con net.Conn, incoming bool) (*Peer, []Endpoint, error) {
```

Comment:
Seems like you dropped all comments on this method. I think this method is critical and complex enough to warrant a short explanation of what it does. I think I'd mention that the concept of handshake was introduced in v10 but that it was built in a backward compatibility way to work with v9 nodes.

Reply:
That was an unintentional oversight due to copy-pasting. Reinstated and made adjustments to include more information.